packages/financial-templates-lib/src/price-feed/utils.js

```diff
@@ -168,6 +168,97 @@ exports.PriceHistory = (getPrice, prices = {}) => {
   };
 };
 
+/**
+ * @notice Constructs the BlockFinder, which is a simple class-like object to find blocks by timestamp, but keeps a
+ *      cache to optimize the search.
+ * @param {Function} requestBlock async function, like web3.eth.getBlock that returns a block for a particular block
+ *      number or returns the latest block if no argument is provided. Blocks returned must have a `number` and
+ */
+exports.BlockFinder = (requestBlock, blocks = []) => {
+  // Grabs the most recent block and caches it.
+  async function getLatestBlock() {
+    const block = await requestBlock();
+    const index = lodash.sortedIndexBy(blocks, block, "number");
+    if (blocks[index]?.number !== block.number) blocks.splice(index, 0, block);
+    return blocks[index];
+  }
+
+  // Grabs the block for a particular number and caches it.
+  async function getBlock(number) {
+    const index = lodash.sortedIndexBy(blocks, { number }, "number");
+    if (blocks[index]?.number === number) return blocks[index]; // Return early if block already exists.
+    const block = await requestBlock(number);
+    blocks.splice(index, 0, block); // A simple insert at index.
+    return block;
+  }
+
+  // Effectively, this is an interpolation search algorithm to minimize block requests.
+  async function findBlock(startBlock, endBlock, timestamp) {
+    // In the case of equality, the endBlock is expected to be passed as the one whose timestamp === the requested
+    // timestamp.
+    if (endBlock.timestamp === timestamp) return endBlock;
+
+    // If there's no equality, but the blocks are adjacent, return the startBlock, since we want the returned block's
+    // timestamp to be <= the requested timestamp.
+    if (endBlock.number === startBlock.number + 1) return startBlock;
+
+    assert(
+      timestamp < endBlock.timestamp && timestamp > startBlock.timestamp,
+      "timestamp not in between start and end blocks"
+    );
+
+    // Interpolating the timestamp we're searching for to block numbers.
+    const totalTimeDifference = endBlock.timestamp - startBlock.timestamp;
+    const totalBlockDistance = endBlock.number - startBlock.number;
+    const blockPercentile = (timestamp - startBlock.timestamp) / totalTimeDifference;
```

Comment:
As mentioned above, if its possible that endBlock == startBlock than this denominator will be 0 and throw unexpectedly

Reply:
This would fail earlier, since the timestamps inequality above would fail (if they are the same block, the timestamp cannot be both above and below the desired timestamp). Need to think a little more about whether they can ever be the same block at the point of entry. My suspicion is that they cannot -- otherwise, I would expect the tests to sporadically fail. I will add a comment about _why_ this cannot happen and add a comment that a precondition of calling this function is that they are not the same block.

This is called from two places:
- getBlockForTimestamp: in this case the blocks are two different indices from the blocks array, so they can't be the same block.
- recursive call: only one of the blocks is new, and the new block that's requested is "clamped" between the previous start and end block. So I don't think they can be the same block.

Reply:
Added a comment making this an explicit precondition.